### PR TITLE
Solver interface for NPPro

### PR DESCRIPTION
**Pull request by ottapav**
NPPro is a commercial solver for strictly convex QP problems. NPPro stands for enhanced Newton Projection with Proportioning method from the literature. Currently, the solver is publically unavailable. But I'd like to ask you to integrate it into the package allowing preliminary evaluations.

### CODE REVIEW

File: qpsolvers/solvers/nppro_.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    -------
+    :
+        Solution returned by the solver.
+
+    Raises
+    ------
+
+    Notes
+    -----
+
```
Comment (stephane-caron) [inline qpsolvers/solvers/nppro_.py:65]: See comment on Notes and keyword arguments below :point_down:

File: qpsolvers/solvers/nppro_.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        # that are not catched by the solver yet
+        warnings.warn(f"NPPro exited with status {exitflag}")
+        return solution
+    solution.x = x
+    if G is not None:
+        solution.z = np.full(h.shape, 0) # not available yet
+    if A is not None:
+        solution.y = np.full(b.shape, 0) # not available yet
+    if lb is not None or ub is not None:
+        solution.z_box = np.full(q.shape, 0) # not available yet
```
Comment (stephane-caron) [inline qpsolvers/solvers/nppro_.py:132]:
```suggestion
        solution.z_box = None  # not available yet
```

File: qpsolvers/solvers/nppro_.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                & A x = b                   \\
+                & lb \leq x \leq ub
+        \end{array}\end{split}
+
+    It is solved using `NPPro`__.
+
+    Parameters
+    ----------
+    P :
+        Symmetric cost matrix.
```
Comment (stephane-caron) [inline qpsolvers/solvers/nppro_.py:173]: If you have assumptions on the cost matrix you can write them down here.

File: qpsolvers/solvers/nppro_.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        Warm-start guess vector.
+    verbose :
+        Set to `True` to print out extra information.
+
+    Returns
+    -------
+    :
+        Solution to the QP, if found, otherwise ``None``.
+
+    Raises
```
Comment (stephane-caron) [inline qpsolvers/solvers/nppro_.py:198]: Feel free to remove Raises if not exception is raised.

File: qpsolvers/solvers/nppro_.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    ------
+
+    Notes
+    -----
+
```
Comment (stephane-caron) [inline qpsolvers/solvers/nppro_.py:203]:
```suggestion
See the Notes section in :func:`nppro_solve_problem`.
```

**Review by stephane-caron — COMMENTED** [review]
LGTM. I have one comment regarding the settings API (to be consistent with the rest of qpsolvers) and doc suggestions o/w.

File: qpsolvers/solvers/nppro_.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    if lb is not None:
+        lb_ = lb
+    if ub is not None:
+        ub_ = ub
+
+    # Create solver object
+    solver = nppro.CreateSolver(n, m)
+
+    # Set options
+    solver.setOption_MaxIter(100);
```
[the remark this answers is not in the document]
Comment (ottapav) [reply qpsolvers/solvers/nppro_.py]: I tried to follow the instance.

File: qpsolvers/solvers/nppro_.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    # Store solution
+    solution = Solution(problem)
+    if exitflag != 0 or np.isnan(x).any():
+        # The second condition typically handle positive semi-definite cases
+        # that are not catched by the solver yet
+        warnings.warn(f"NPPro exited with status {exitflag}")
+        return solution
+    solution.x = x
+    if G is not None:
+        solution.z = np.full(h.shape, 0) # not available yet
```
[the remark this answers is not in the document]
Comment (ottapav) [reply qpsolvers/solvers/nppro_.py]: Fixed. Noted.

File: qpsolvers/solvers/nppro_.py (hunk 3 above)
Comment (ottapav) [reply qpsolvers/solvers/nppro_.py]: assumption mentioned there is P: Positive definite cost matrix

**Comment by stephane-caron** [comment]
LGTM!

Just to check: NPPro does not provide a verbose option, correct?

**Comment by ottapav** [comment]
> LGTM! I'm in the process of merging your contribution.
> 
> Just to check: there is no verbose option in NPPro, correct? (Otherwise we can use the `verbose` argument which is currently unused.)

Correct. It is primarily developed for embedded devices and currently, no such feature is available yet.

**Comment by stephane-caron** [comment]
Merged. Thanks for contributing this interface.

**Pull request merged**